### DISCUSSION TITLE
How do implement a "hook" to take screenshot on Error automatically

### DISCUSSION THREAD

**Discussion by lehmannk** [fzyzcjy/flutter_convenient_test#374] [Q&A]
Currently, convenience test does not take a log screenshot when e.g. assertion or execution errors occur. Now, instead of capsulating everything with try - catch blocks etc, is there a convenient way to let convenience test automatically create a log screenshot in these cases, e.g. through `FlutterError.onError`
<img width="1707" alt="image" src="https://github.com/fzyzcjy/flutter_convenient_test/assets/36925534/d021c6dc-579b-43ab-a992-1c265467cbf5">

**Comment by fzyzcjy** [comment]
Firstly, is there screenshot for the item *above* the error?

Secondly, currently there are some snapshots when some kind of errors happen, e.g. https://github.com/fzyzcjy/flutter_convenient_test/blob/2dfee8760b820e3951de7dceb4edc96c14cc6980/packages/convenient_test_dev/lib/src/functions/find.dart#L281.

**Reply by lehmannk** [reply]
sorry for the late reply. Yes, there are screenshots for the items above but they do not pertain to the erroneous situation. Another example is
<img width="652" alt="image" src="https://github.com/fzyzcjy/flutter_convenient_test/assets/36925534/47f831a9-e52d-416b-bdea-22eea8e22d08">
where there is a screenshot for the DRAG entry (initial state before drag) but no screenshot at the situation when the error was fired (after drag)

So, do you suggest to add more log.snapshot() calls to convenience_test as a PR contribution or rather solve it with local try-catch in the test code?

**Reply by lehmannk** [reply]
Ah, I think I see the problem. While convenience test extends all "native" finders and commands in https://github.com/fzyzcjy/flutter_convenient_test/blob/2dfee8760b820e3951de7dceb4edc96c14cc6980/packages/convenient_test_dev/lib/src/functions/find.dart

I am using PatrolFinder instead to locate and operate controls (because these are way easier to use). For PatrolFinder there is no extension in convenience test. So would you suggest copying convenience test`s find.dart
and adapting everything for PatrolFinder or is there an easier way to integrate another 3rd party finder lib?

**Reply by fzyzcjy** [reply]
Hmm, good question! I agree there needs to be some code that glues the two together. As for easier adaption, maybe we can abstract away the common parts a little bit.

**Reply by lehmannk** [reply]
As a first step I copied convenience test`s find.dart and created these new extensions
``` 
extension PatrolFinderExtension on PatrolFinder {
...
Future<void> tap({SettlePolicy? settlePolicy, Duration? visibleTimeout, Duration? settleTimeout}) =>
      PatrolFinderCommandExtension.auto(this)
          .tap(settlePolicy: settlePolicy, visibleTimeout: visibleTimeout, settleTimeout: settleTimeout);
...
}
```
with 
```
class PatrolFinderCommandExtension extends TCommand {
...
 Future<void> tap({
    SettlePolicy? settlePolicy,
    Duration? visibleTimeout,
    Duration? settleTimeout,
  }) =>
      act(
        act: (log) => tap(settlePolicy: settlePolicy, visibleTimeout: visibleTimeout, settleTimeout: settleTimeout),
        preCondition: ElementHitTestableMatcher(t.tester),
        logTitle: 'TAP',
        logMessage: finder.describeMatch(Plurality.one),
        settle: settleTimeout != null,
      );
...
}
```
where the `tap` methode references PatrolFinder's `tap`method. But my test code with the PatrolFinder `tap()` method does not run into this extension. What else is required to connect the extension with convenience test runs?

**Reply by fzyzcjy** [reply]
> But my test code with the PatrolFinder tap() method does not run into this extension

A first debugging may be, click through all the functions and see which one calls which one (e.g. ctrl-click, or any hotkey you use). For example, a quick look seems that the `(log) => tap(...)` may be a recursive call (surely I can be wrong - since I look at it by eyes instead of using an IDE)

**Reply by lehmannk** [reply]
Ok, I think I found a way. I need to extend both `PatrolFinder` (in order to map PatrolFinder to ConvenienceTest finders) as well as `TFinderCommand` in order to actually implement PatrolFinder commands instead of ConvenienceTest commands such as `tap()` etc... A bit fiddly but it works :-)

**Comment by lehmannk** [comment]
Hi,

the latest PR <https://github.com/fzyzcjy/flutter_convenient_test/pull/426> has not triggered a new release yet. Can you do that?

Greets from Berlin :-)


Kai

**Reply by fzyzcjy** [reply]
Definitely! But I am too busy this several days, will do it ASAP

**Reply by fzyzcjy** [reply]
done
